src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs

```diff
@@ -1018,71 +1025,81 @@ private async Task SendWithLock(ConnectionState expectedConnectionState, HubMess
                 {
-                    resultException = ex;
+                    var task = handler.InvokeAsync(invocation.Arguments);
+                    if (handler.HasResult && task is Task<object?> resultTask)
+                    {
+                        result = await resultTask.ConfigureAwait(false);
+                        hasResult = true;
+                    }
+                    else
+                    {
+                        await task.ConfigureAwait(false);
+                    }
+                }
+                catch (Exception ex)
+                {
+                    Log.ErrorInvokingClientSideMethod(_logger, invocation.Target, ex);
+                    if (handler.HasResult)
+                    {
+                        resultException = ex;
+                    }
                 }
             }
-        }
 
-        if (expectsResult)
-        {
-            if (resultException is not null)
+            if (expectsResult)
             {
-                await SendWithLock(connectionState, CompletionMessage.WithError(invocation.InvocationId!, resultException.Message), cancellationToken: default).ConfigureAwait(false);
+                if (resultException is not null)
+                {
+                    await SendWithLock(connectionState, CompletionMessage.WithError(invocation.InvocationId!, resultException.Message), cancellationToken: default).ConfigureAwait(false);
+                }
+                else if (hasResult)
+                {
+                    await SendWithLock(connectionState, CompletionMessage.WithResult(invocation.InvocationId!, result), cancellationToken: default).ConfigureAwait(false);
+                }
+                else
+                {
+                    Log.MissingResultHandler(_logger, invocation.Target);
+                    await SendWithLock(connectionState, CompletionMessage.WithError(invocation.InvocationId!, "Client didn't provide a result."), cancellationToken: default).ConfigureAwait(false);
+                }
             }
             else if (hasResult)
             {
-                await SendWithLock(connectionState, CompletionMessage.WithResult(invocation.InvocationId!, result), cancellationToken: default).ConfigureAwait(false);
-            }
-            else
-            {
-                Log.MissingResultHandler(_logger, invocation.Target);
-                await SendWithLock(connectionState, CompletionMessage.WithError(invocation.InvocationId!, "Client didn't provide a result."), cancellationToken: default).ConfigureAwait(false);
+                Log.ResultNotExpected(_logger, invocation.Target);
             }
         }
-        else if (hasResult)
+        catch (Exception ex)
         {
-            Log.ResultNotExpected(_logger, invocation.Target);
+            // Avoid unobserved task exceptions
+            // Currently the only exceptions that should be able to occur are if a client result completion message is being sent and fails
```

Comment:
If this is true, why do we have this massive try/catch? Can't we limit this to just the `if (expectsResult)` branches?

Also, is "Error processing invocation '{InvocationId}'." at trace level really descriptive enough? Can't we just log that the connection closed before the client could produce a result?

I think this should log it at the error level similar to ErrorInvokingClientSideMethod and say "Error returning result for invocation '{InvocationId}' to '{Target}' because the underlying connection closed."  At least make it a warning.

Also, I think there's a very good argument that there shouldn't be an error log for `ErrorInvokingClientSideMethod` if we've successfully sent back an errored completion message to the server.